core-java-modules/core-java-concurrency-advanced-4/src/main/java/com/baeldung/producerconsumer/Consumer.java

```diff
@@ -16,21 +16,24 @@ public void run() {
 
     public void consume() {
         while (runFlag) {
-            Message message;
-            if (dataQueue.isEmpty()) {
-                try {
-                    dataQueue.waitOnEmpty();
-                } catch (InterruptedException e) {
-                    e.printStackTrace();
+            synchronized (this) {
+                if (dataQueue.isEmpty()) {
+                    try {
+                        dataQueue.waitOnEmpty();
+                    } catch (InterruptedException e) {
+                        e.printStackTrace();
+                        break;
+                    }
+                }
+                if (!runFlag) {
                     break;
                 }
+                if (!dataQueue.isEmpty()) {
```

Comment:
Since we change line 20 to "while", you can remove this check

Reply:
Done